rfcs/0000-unified-configuration.ftd

```diff
@@ -0,0 +1,188 @@
+
+-- lib.rfc: RFC: Unified Configuration
+id: unified-config
+status: proposal
+
+TODO: Write a brief summary of the RFC here.
+
+
+-- lib.motivation:
+
+;; TODO: add links to the following paragraph
+Currently, `fastn` supports configuration through environment variables, and in
+some way, through the `FASTN.ftd` file. Environment variables are good to
+change the behaviour of the cli or to pass secret information but it's not
+ideal for defining complex configuration for a framework like `fastn` that
+provides various configurable features like processors, oauth authentication
+and, apps.
+
+;; TODO: I can do better than this
+This RFC proposes a way of dealing with configuration that is centralised and
+leveraged the existing ability of the `fastn` cli to read environment
+variables.
+
+-- end: lib.motivation
+
+
+-- lib.detailed-design:
+
+-- ds.h2: `fastn/config` - A virtual module to assist configuration definition.
+
+Add a virtual module (`fastn/config`) similar to existing [assets](TODO: add
+link) module.
+
+;; TODO add footer section for explanors like foreign variables and virtual module
+- This module contains all special configuration types defined by the fastn
+  framework. This module also provides safe access to environment variables
+  through foreign variables.
+
+- `FASTN.ftd` file will be central to the whole configuration system. Users can
+  modify the behaviour of a dependency by passing variables when they're added.
+  Example:
+
+-- ds.code:
+lang: ftd
+
+;; FASTN.ftd of test-todos
+\-- import: fastn
+
+\-- fastn.package: test-todos
+
+\;; fastn.app works in similar way
+\-- fastn.dependency: amitu.com/todos
+max-todo-per-page: 30
+ignore-value: test \;; amitu.com/todos doesn't have this variable so we ignore this
+
+-- end: ds.code
+
+
+-- ds.markdown:
+
+The above snippet is configuring `amitu.com/todos` by modifying variables that
+are in its `FASTN.ftd` file.
+
+The `FASTN.ftd` of `amitu.com/todos` will look like this:
+
+-- ds.code:
+lang: ftd
+
+;; FASTN.ftd of amitu.com/todos
+\-- import: fastn
+
+\-- fastn.package: amitu.com/todos
+
+\-- integer max-todo-per-page: 10 \;; becomes 30 when used by test-todos
+
+\-- string sample-title: TODO App \;; this was not modified by test-todos so this stays intact
+
+-- end: ds.code
+
+
+-- ds.markdown:
+
+- While adding dependency, users can modify any variable defined in the
+  `FASTN.ftd` file of the said dependency. Any variable that is not defined
+  there is simply ignored
+
+- Variables defined in `FASTN.ftd` are also considered by the framework
+  features. Example: Github OAuth feature (this is a processor) can specify
+  that it needs `gh-oauth` variable of type `github` defined in user's
+  `FASTN.ftd` file for it to work in user's `FASTN.ftd` file for it to work
+  (the error handling when `gh-oauth` is not available depends on the
+  implementation).
+
+- Variables created using types from `fastn/congi` (as well as primitive types)
```

Comment:
```suggestion
- Variables created using types from `fastn/config` (as well as primitive types)
```

Reply:
[resolved in this commit](https://github.com/fastn-stack/fastn.com/pull/523/commits/5fcd2fd435f4479915c450606a4c78f19e9e4a61)